### PR TITLE
Add infrastructure to migrate bpmn element processors

### LINKED ISSUE #6350: Add infrastructure to support migrating the bpmn element processors
**Description**

In order to migrate the BPMN Element processors (#6189, #6172, etc.) we need additional infrastructure to support the migration.

### PR DESCRIPTION
**Pull request by korthout**
## Description

The BPMN element processors require extra infrastructure before they can be migrated. To reduce the size of the individual processor migration PRs, this PR adds that code separately.

The largest change here is to transition the referred element's state when processing one of the new commands: so transition to Element_Activating when processing Activate_Element. This helps the processor to receive a context that actually resembles the current state, instead of the intended state.

This PR also replaces the usage of the `StateWriter` to the `EventWriter` for the `CreateWorkflowInstanceProcessor`, because this would result in problems. For a full description see the specific commit message.

For the reviewer: First, I think we shouldn't mind the sonar failure. It's mainly concerns the BpmnStreamProcessor's long switch statement with large cases, which will be reduced in #6202 anyways. Secondly, do you think we need any tests for this? This does not change any real behaviour, since none of the element processors have been migrated yet.

## Related issues

closes #6350 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/WorkflowInstanceStateTransitionGuard.java [hunk 1]
```diff
@@ -115,19 +122,31 @@ public boolean isValidStateTransition(final BpmnElementContext context) {
...
-              expectedState, currentState));
-
+              "Expected element instance to be in state '%s' or one of '%s' but was '%s'.",
+              expectedState, Arrays.toString(otherExpected), currentState));
     } else {
       return Either.right(elementInstance);
     }
   }
 
+  private Either<String, ?> hasElementInstanceWithState(
```
Comment (pihme) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/WorkflowInstanceStateTransitionGuard.java:139]: Do we need both methods? (`hasElementInstanceWithState/inState`) If one is supposed to be a shortcut to call the other, maybe we can always use the shortcut and inline the called method

**Review by pihme — APPROVED** [review]
Looks good to me. Just two comments to consider.

On the tests. I don't really need tests, tbh. While reviewing it I was thinking of  a property based test which verifies that for every value of the intent enum, it is either a command or an event but never both. Then if somebody adds a new intent and forgets to put it into either of the lists the test would fail. This should only happen very rarely, though. So the test isn't adding very much.

On the Sonar side. I agree we can ignore the warnings. But maybe we could be a little more explicit about which warnings we want to ignore: https://stackoverflow.com/questions/10971968/turning-sonar-off-for-certain-code

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/WorkflowInstanceStateTransitionGuard.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/WorkflowInstanceStateTransitionGuard.java:139]: I've tried this, but it looks worse to me.

**Comment by korthout** [comment]
@pihme please have another look

**Review by pihme — APPROVED** [review]
lgtm 🎉

**Comment by korthout** [comment]
bors merge

**Pull request merged**